### PR TITLE
Made editor compound balance assume ATP equilibrium

### PR DESCRIPTION
**Pull request by Leinourdian**
**Brief Description of What This PR Does**

This adds a new method that calculates compound balance assuming that the cell will produce at most as much ATP as it consumes. The editor uses the new method.

Notes:
-	All in-game text is kept the same, so this assumes that players will find this change intuitive.
-	Initial compounds still use the old method for compound balance. Changing them to use the new one might make sense.
-	I’m not sure, but this might be slightly different from true compound balance during gameplay if e.g. process ordering and storage capacity make different processes happen at different speeds.
-	The code of the new method `CalculateCompoundBalanceAtEquilibrium` is very similar to the old method `CalculateCompoundBalance`. If wanted, the old method could be removed, or the methods or their parts could be merged into a single method.
-	This method should be pretty simple to adapt to use different presets like not moving or lacking certain compounds. Could be useful for #2068 in the future.
-	This doesn’t show ATP. If it did, it would show either 0 or a negative value. I think it’s probably unnecessary to show ATP since the ATP balance bar exists.
-	If you consume more ATP than you produce, the method treats that as you being at equilibrium. This means that ATP consuming processes (like producing toxins) are always considered to run at full speed.

**Related Issues**

Closes [#2069](https://github.com/Revolutionary-Games/Thrive/issues/2069)

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/microbe_stage/ProcessSystem.cs [hunk 1]
```diff
@@ -165,6 +165,11 @@ public ProcessSystem(Node worldRoot)
     ///   Computes the compound balances for given organelle list in a patch and at a given time during the day (or
     ///   using longer timespan values)
     /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     Ignores how much ATP the cell needs
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSystem.cs:170]:
```suggestion
    ///     Ignores how much ATP the cell needs (i.e. assumes instead that all processes run always at max possible speed)
```

File: src/microbe_stage/ProcessSystem.cs [hunk 2]
```diff
@@ -207,6 +212,77 @@ void MakeSureResultExists(Compound compound)
...
+    ///   Computes the compound balances for given organelle list in a patch and at a given time during the day (or
+    ///   using longer timespan values)
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     Assumes that the cell produces at most as much ATP as it consumes
+    ///   </para>
+    /// </remarks>
+    public static Dictionary<Compound, CompoundBalance> ComputeCompoundBalanceAtEquilibrium(
+        IEnumerable<OrganelleTemplate> organelles, BiomeConditions biome, CompoundAmountType amountType,
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSystem.cs:225]: Is there a reason why this takes in organelle templates and not organelle definitions like the `ComputeCompoundBalance` method?

File: src/microbe_stage/ProcessSystem.cs [hunk 3]
```diff
@@ -207,6 +212,77 @@ void MakeSureResultExists(Compound compound)
...
+        foreach (var organelle in organelles)
+        {
+            foreach (var process in organelle.Definition.RunnableProcesses)
+            {
+                var speedAdjusted = CalculateProcessMaximumSpeed(process, biome, amountType);
+
+                useRatio = false;
+
+                // If the cell produces more ATP than it needs, its ATP producing processes need to be toned down
+                if (speedAdjusted.Outputs.Select(o => o.Key).Contains(ATP) && consumptionProductionRatio < 1.0f)
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSystem.cs:250]:
```suggestion
                if (speedAdjusted.Outputs.ContainsKey(ATP) && consumptionProductionRatio < 1.0f)
```
Would that work?

**Comment by hhyyrylainen** [comment]
There's just a tiny bit of changes needed for this so this should be completable in time for the RC on Saturday.

File: src/microbe_stage/ProcessSystem.cs (hunk 2 above)
Comment (Leinourdian) [reply src/microbe_stage/ProcessSystem.cs:225]: Not really actually, I can change this to be consistent with the other method.

File: src/microbe_stage/ProcessSystem.cs (hunk 3 above)
Comment (Leinourdian) [reply src/microbe_stage/ProcessSystem.cs:250]: Unfortunately not, `Outputs` doesn't have the `ContainsKey` method or anything similar as far as I can tell.

Comment (hhyyrylainen) [reply src/microbe_stage/ProcessSystem.cs:250]: I think refactoring the outputs property to be like this: `public IReadOnlyDictionary<Compound, float> Outputs => WritableOutputs;` would be a good choice here.

**Review by hhyyrylainen — APPROVED** [review]
The code change looks fine to me now. You Just need to resolve the one merge conflict before this can be merged. And also I'll give time at least until tomorrow for playtesting this, in case someone does it.

**Comment by buckly90** [comment]
Works wonderfully and without any visible error. Outstanding job!

**Pull request merged**